File: PAM/communication/frame.c
#include <stddef.h>
#include <sys/types.h>

#include "frame.h"

static void put_u32_be(uint8_t *dst, uint32_t value) {
    dst[0] = (uint8_t) (value >> 24);
    dst[1] = (uint8_t) (value >> 16);
    dst[2] = (uint8_t) (value >> 8);
    dst[3] = (uint8_t) value;
}

static uint32_t get_u32_be(const uint8_t *src) {
    return ((uint32_t) src[0] << 24) |
           ((uint32_t) src[1] << 16) |
           ((uint32_t) src[2] << 8) |
           (uint32_t) src[3];
}

ssize_t frame_encode(Message message, uint8_t *out, size_t out_cap) {
    size_t payload_len = (size_t) message.payload_len;
    size_t frame_length = FRAME_HEADER_LEN + payload_len;

    if (out_cap < frame_length) return -1;

    out[0] = FRAME_MAGIC0;
    out[1] = FRAME_MAGIC1;
    out[2] = FRAME_VERSION;
    out[3] = (uint8_t) message.type;
    put_u32_be(out + 4, message.seq);
    put_u32_be(out + 8, message.payload_len);

    for (size_t i = 0; i < payload_len; i++) {
        out[FRAME_HEADER_LEN + i] = message.payload[i];
    }

    return (ssize_t) frame_length;
}

int frame_decode(const uint8_t *in, size_t in_len, Message *out, size_t payload_cap) {
    if (in_len < FRAME_HEADER_LEN) return FRAME_TOO_SMALL;
    if (in[0] != FRAME_MAGIC0 || in[1] != FRAME_MAGIC1) return INVALID_MAGIC_BITS;
    if (in[2] != FRAME_VERSION) return VERSION_MISMATCH;

    size_t payload_len = (size_t) get_u32_be(in + 8);
    if (payload_len > in_len - FRAME_HEADER_LEN) return OUT_TOO_SMALL;
    if (payload_len > payload_cap) return PAYLOAD_TOO_BIG;

    out->type = (MsgType) in[3];
    out->seq = get_u32_be(in + 4);
    out->payload_len = (uint32_t) payload_len;
    out->payload = in + FRAME_HEADER_LEN;
    return 0;
}

File: App/app/src/main/java/uk/ac/york/emrys/Frame.kt
package uk.ac.york.emrys

sealed class FrameCodecException(message: String) : IllegalArgumentException(message) {
    class FrameTooSmall : FrameCodecException("Frame is too small")
    class InvalidMagicBits : FrameCodecException("Invalid magic bits")
    class VersionMismatch : FrameCodecException("Version mismatch")
    class OutputTooSmall : FrameCodecException("Output buffer is too small")
    class PayloadTooBig : FrameCodecException("Payload is too big")
    class InvalidMessageType : FrameCodecException("Invalid message type")
    class PayloadLengthMismatch : FrameCodecException("Payload length does not match payload data")
}

class Frame {
    companion object {
        private const val FRAME_HEADER_LEN = 12
        private const val FRAME_MAGIC0: Byte = 0x53  // 'S'
        private const val FRAME_MAGIC1: Byte = 0x50  // 'P'
        private const val FRAME_VERSION: Byte = 0x01

        private fun putU32Be(dst: ByteArray, offset: Int, value: UInt) {
            dst[offset] = ((value shr 24) and 0xFFu).toByte()
            dst[offset + 1] = ((value shr 16) and 0xFFu).toByte()
            dst[offset + 2] = ((value shr 8) and 0xFFu).toByte()
            dst[offset + 3] = (value and 0xFFu).toByte()
        }

        private fun getU32Be(src: ByteArray, offset: Int): UInt {
            return ((src[offset].toUInt() and 0xFFu) shl 24) or
                    ((src[offset + 1].toUInt() and 0xFFu) shl 16) or
                    ((src[offset + 2].toUInt() and 0xFFu) shl 8) or
                    (src[offset + 3].toUInt() and 0xFFu)
        }

        private fun UInt.toIntOrThrow() =
            if (this <= Int.MAX_VALUE.toUInt()) toInt() else throw FrameCodecException.PayloadTooBig()
    }

    fun frameEncoder(message: Message, outputCapacity: Int = Int.MAX_VALUE): ByteArray {
        val payloadLength = message.payloadLen.toIntOrThrow()
        if (payloadLength > message.payload.size) {
            throw FrameCodecException.PayloadLengthMismatch()
        }

        val frameLength = FRAME_HEADER_LEN + payloadLength
        if (outputCapacity < frameLength) {
            throw FrameCodecException.OutputTooSmall()
        }

        val frame = ByteArray(frameLength)
        frame[0] = FRAME_MAGIC0
        frame[1] = FRAME_MAGIC1
        frame[2] = FRAME_VERSION
        frame[3] = message.type.value
        putU32Be(frame, 4, message.seq)
        putU32Be(frame, 8, message.payloadLen)

        if (payloadLength > 0) {
            System.arraycopy(message.payload, 0, frame, FRAME_HEADER_LEN, payloadLength)
        }

        return frame
    }

    fun frameDecoder(frame: ByteArray, payloadCapacity: Int = Int.MAX_VALUE): Message {
        if (frame.size < FRAME_HEADER_LEN) throw FrameCodecException.FrameTooSmall()
        if (frame[0] != FRAME_MAGIC0 || frame[1] != FRAME_MAGIC1) {
            throw FrameCodecException.InvalidMagicBits()
        }
        if (frame[2] != FRAME_VERSION) throw FrameCodecException.VersionMismatch()

        val type = MsgType.fromByte(frame[3])
            ?: throw FrameCodecException.InvalidMessageType()

        val seq = getU32Be(frame, 4)
        val payloadLen = getU32Be(frame, 8)
        val payloadLenInt = payloadLen.toIntOrThrow()

        if (payloadLenInt > frame.size - FRAME_HEADER_LEN) {
            throw FrameCodecException.OutputTooSmall()
        }
        if (payloadLenInt > payloadCapacity) {
            throw FrameCodecException.PayloadTooBig()
        }

        val payload = frame.copyOfRange(FRAME_HEADER_LEN, FRAME_HEADER_LEN + payloadLenInt)

        return Message(type, seq, payload, payloadLen)
    }
}
